File: TPII-2025-2/PadraoGoF/Estrutural/Decorator/Decorator01.js
//Componente notificador
class ComponenteNotificador{
    enviar(msg){
        console.log(`Enviando msg: ${msg}`);
    }
}

//Componente decorator base
class DecoratorNotificador{
    constructor(componente){
        this.componente = componente;
    }

    enviar(msg){
        this.componente.enviar(msg);
    }
}

//Componentes decorator concreto 1 - SMS
class Notificador extends DecoratorNotificador{
    enviar(msg){
        super.enviar(msg);
        console.log(`Enviando msg por SMS: ${msg}`);
    }
}

// Componentes Decorador Concreto 1 - SMS:
class NotificadorSMS extends DecoratorNotificador{
    enviar(msg){
        super.enviar(msg);
        console.log(`Enviando msg por SMS: ${msg}`);
    }
}

// Componentes decorador concreto 2 - wpp
class NotificaWhatsapp extends DecoratorNotificador{
    enviar(msg){
        super.enviar(msg);
        console.log(`Enviando msg por wpp: ${msg}`);
    }
}

// Componentes decorador concreto 3 - email
class NotificaEmail extends DecoratorNotificador{
    enviar(msg){
        super.enviar(msg);
        console.log(`Enviando msg por email: ${msg}`);
    }
}

// Uso pelo cliente
const notificador = new ComponenteNotificador();
const notificadorSMS = new NotificadorSMS(notificador);
const notificadorWpp = new NotificaWhatsapp(notificadorSMS);

notificadorWpp.enviar("Pedido saindo");

File: TPII-2025-2/PadraoGoF/Estrutural/Decorator/Decorator01.py
# Componente Notificador:
class ComponenteNotificador:
    def enviar(self, msg):
        print(f'Enviado msg: {msg}')

# Componente decorator base
class DecoradorNotificador:
    def __init__(self,componente):
        self.componente = componente

    def enviar(self, msg):
        self.componente.enviar(msg)

# Componente decorator concretro 1 - sms
class NotificadorSMS(DecoradorNotificador):
    def enviar(self, msg):
        super().enviar(msg)
        print(f'Enviando Msg por SMS: {msg}')

#Componente decorator concreto 2 - wpp
class NotificadorWhatsapp(DecoradorNotificador):
    def enviar(self, msg):
        super().enviar(msg)
        print(f'Enviando Msg por wpp: {msg}')

#Componente decorator concreto 3 - email
class NotificacaoEmail(DecoradorNotificador):
    def enviar(self, msg):
        super().enviar(msg)
        print(f'Enviando Msg por email: {msg}')

notificador = ComponenteNotificador()
notificadorSMS = NotificadorSMS(notificador)
notificadorWhatsapp = NotificadorWhatsapp(notificadorSMS)

notificadorWhatsapp.enviar("Pedido saindo...")